### PR TITLE
[core] - refactor: fix MistralChatMessage serialization

### PR DESCRIPTION
**Pull request by JulesBelveze**
## Description

This PR aims at fixing the following issue encountered with Mistral Large:
```
[{"value":null,"error":"Error querying `mistral:mistral-large-latest`: error=[model_error(retryable=, request_id=...)] \"Assistant message must have either content or tool_calls, but not both.\"","meta":null}]
```

It introduces the following changes:
 - Serialize `content` in `MistralChatMessage` only when there are no `tool_calls` present
 - Ensure `AssistantChatMessage` creation aligns with conditional presence of `tool_calls` or `content`
 - Strip `content` from `AssistantChatMessage` if `tool_calls` are present to prevent unnecessary data
 - Adjust `MistralAILLM` to clear `content` when `tool_calls` are set to maintain data consistency

**References:**
- https://github.com/dust-tt/tasks/issues/1101

## Risk

Breaking/unexpected interactions with Mistral

## Deploy Plan

- Deploy `core`

### CODE REVIEW

**Comment by JulesBelveze** [comment]
@fontanierh pardon my Rust, looking for feedback 😇

File: core/src/providers/mistral.rs [hunk 1]
```diff
@@ -148,19 +149,31 @@ impl TryFrom<&ChatMessage> for MistralChatMessage {
...
-                    Some(fc) => Some(
-                        fc.iter()
-                            .map(|f| MistralToolCall::try_from(f))
-                            .collect::<Result<Vec<_>>>()?,
-                    ),
-                    None => None,
-                },
-                tool_call_id: None,
-            }),
+            ChatMessage::Assistant(assistant_msg) => {
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:152]: could do something like 

```
 Ok(MistralChatMessage {
                        role: MistralChatMessageRole::Assistant,
                        content: match assistant_msg.function_calls {
                            Some(_) -> None,
                            None => assistant_msg.content
                         },
                        tool_calls: assistant_msg
                            .function_calls
                            .as_ref()
                            .map(|fc| {
                                fc.iter()
                                    .map(|f| MistralToolCall::try_from(f))
                                    .collect::<Result<Vec<_>>>()
                            })
                            .transpose()?,
                        tool_call_id: None,
                })
                    ```

Comment (fontanierh) [reply core/src/providers/mistral.rs:152]: (to avid the slight repetition and if/else)

File: core/src/providers/mistral.rs [hunk 2]
```diff
@@ -216,31 +229,18 @@ impl TryFrom<&MistralChatMessage> for AssistantChatMessage {
...
         let role = ChatMessageRole::from(cm.role.clone());
-        let content = match cm.content.as_ref() {
-            Some(c) => Some(c.clone()),
-            None => None,
-        };
 
-        let function_calls = if let Some(tool_calls) = cm.tool_calls.as_ref() {
-            let cfc = tool_calls
-                .into_iter()
+        let (content, function_calls) = if let Some(tool_calls) = cm.tool_calls.as_ref() {
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:233]: here I would do a match as `let (...) = if let ... = ...` is pretty cumbersome

Comment (fontanierh) [reply core/src/providers/mistral.rs:233]: i.e:

```
let (content, function_calls) = match cm.tool_calls.as_ref() {
   None => (cm.content.clone(), None)
   Some(tool_calls) => (None, tool_calls.iter()...)
}

**Review by fontanierh — COMMENTED** [review]
just a few comments !

Also are you sure all the clones are necessary ? It's hard to tell without actually pulling the branch so maybe there are but worth checking

**Comment by JulesBelveze** [comment]
@fontanierh thanks for feedback 🙏🏼 
Double checked and can't get rid of any `clone` 😥

File: core/src/providers/mistral.rs [hunk 3]
```diff
@@ -150,15 +151,19 @@ impl TryFrom<&ChatMessage> for MistralChatMessage {
         match cm {
             ChatMessage::Assistant(assistant_msg) => Ok(MistralChatMessage {
                 role: MistralChatMessageRole::Assistant,
-                content: assistant_msg.content.clone(),
-                tool_calls: match assistant_msg.function_calls.as_ref() {
-                    Some(fc) => Some(
+                content: match assistant_msg.clone().function_calls {
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:154]: are we sure we need this clone ? seems a bit weird

File: core/src/providers/mistral.rs [hunk 4]
```diff
@@ -150,15 +151,19 @@ impl TryFrom<&ChatMessage> for MistralChatMessage {
...
+                    Some(_) => None,
+                    None => assistant_msg.clone().content,
```
Comment (fontanierh) [inline core/src/providers/mistral.rs:156]: can we clone just content instead of `assistant_msg` ?

**Review by fontanierh — APPROVED** [review]
LGTM just a small q regarding clone

File: core/src/providers/mistral.rs (hunk 3 above)
Comment (JulesBelveze) [reply core/src/providers/mistral.rs]: Got rid of it 👌🏼

**Pull request merged**